### ISSUE TITLE
Pattern matching

### ISSUE THREAD

**Issue by gcanti** [gcanti/tcomb#121]
On gitter @jgoux (thanks) has raised an interesting topic: bringing some kind of pattern matching to tcomb. 
Something along the lines of the nice @paldepind's library https://github.com/paldepind/union-type.

The first sketch of the API seems nice (especially with ES6 syntax):

```js
const A = t.struct(...);
const B = t.struct(...);

t.match(x, {
  A: (x) => 'doSomething',
  B: (x) => 'doSomethingElse',
  _: (x) => 'optionalCatchAllClause' // optional
})
```

My main concern is that relying on the (optional) name of a type is kind of brittle. 
I'd prefer matching on type constructors (it's also more general):

```js
t.match(x, [
  [ t.String, (s) => 'a string' ],
  [ t.Number, (n) => n > 2, (n) => 'a number gt 2' ], // case with a guard (optional)
  [ t.Number, (n) => 'a number lte 2' ],
  [ A, (a) => 'an instance of A' ],
  [ t.Any, (x) => 'other...'] // catch all
]);
```

**Comment by jgoux** [comment]
:+1: for the second approach.

**Comment by volkanunsal** [comment]
A third alternative you might consider is this:

```javascript
t.match(x, 
  A: (x) => 'doSomething',
  B: (x) => 'doSomethingElse',
  _: (x) => 'optionalCatchAllClause' // optional
)
```

This is probably the simplest one. You have intrinsic ordering in the `arguments`, so there is no need for another array.

**Comment by gcanti** [comment]
Hi @volkanunsal, thanks for your feedback. Stretching your suggestion and using a strong convention one could even remove all the arrays:

```js
t.match(x,
  t.String, (s) => 'a string',
  t.Number, (n) => n > 2, (n) => 'a number gt 2', // clause with a guard (optional)
  t.Number, (n) => 'a number lte 2',
  A, (a) => 'an instance of A',
  t.Any, (x) => 'other...' // catch all
);
```

- intrinsic ordering in the `arguments`
- clauses are always separated by a type

**Comment by volkanunsal** [comment]
This is the best one so far. :+1:

**Comment by jgoux** [comment]
I agree, the last proposition is even better. Great job !

**Comment by volkanunsal** [comment]
I'm curious why you even need a function in the matchers. Don't the function parameters all refer to `x` anyway? Can't you write it this way to make it even simpler?

```javascript
t.match(x,
  t.String, 'a string',
  t.Number, ['a number gt 2', x > 2], // case with a guard (optional)
  t.Number, 'a number lte 2',
  A, 'an instance of A',
  t.Any, 'other...' // catch all
);
```

**Comment by gcanti** [comment]
> why you even need a function in the matchers

In short for:

- reuse (first case)
- correctness: some expression could be invalid or throw an error (second case)
- lazyness: expensive operations are performed only when needed (third case)

```js
function isEven(x) {
  return x % 2 === 0;
}

function f(x) {
  return t.match(x,
    t.Number, isEven/* reuse */, () => 'a number',
    t.Number, (n) => /* a string * 2 makes no sense */ n * 2,
    t.String, () => /* say this is an expensive operation */ 'a string'
  );
}

console.log(f('a'));
```

**Comment by volkanunsal** [comment]
Ah, yes, good point.

**Comment by jgoux** [comment]
I was wondering, does it work with union of union types ?
If the type to match is a Union type, does it recursively check the types to check all the possibility ?
I want to use this to be able to do command routing/subrouting
Example : 
```js
// An aggregate can handle a predefined list of commands
const TodoListAggregate.Command = union([AddTodo, EditTodo])
// An aggregate route the command to its handler
const TodoListAggregate.handle = (command) => {
    return match(command,
        AddTodo, (c) => TodoListAggregate.addTodo(c),
        EditTodo, (c) => TodoListAggregate.editTodo(c)
}

// The domain accept any command defined in the aggregates
const Command = union([TodoListAggregate.Command, AnotherAggregate.Command])
// The domain route the command to its aggregate
const handle = (command) => {
    return match(command,
        TodoListAggregate.Command, (c) => TodoListAggregate.handle(c),
        AnotherAggregate.Command, (c) => AnotherAggregate.handle(c)
}
```

**Comment by gcanti** [comment]
@jgoux the current implementation (merged in the master branch) uses `Union.is` so yes is recursive:

```js
var A = t.subtype(t.String, function (s) { return s === 'A'; });
var B = t.subtype(t.String, function (s) { return s === 'B'; });
var C = t.subtype(t.String, function (s) { return s === 'C'; });
var D = t.subtype(t.String, function (s) { return s === 'D'; });
var E = t.subtype(t.String, function (s) { return s === 'E'; });
var F = t.subtype(t.String, function (s) { return s === 'F'; });
var G = t.subtype(t.String, function (s) { return s === 'G'; });
var H = t.subtype(t.String, function (s) { return s === 'H'; });

var U1 = t.union([A, B]);
var U2 = t.union([C, D]);
var U3 = t.union([E, F]);
var U4 = t.union([G, H]);

var UU1 = t.union([U1, U2]);
var UU2 = t.union([U3, U4]);

t.match('F',
  UU1, function () { return '1'; },
  UU2, function () { return '2'; }
); // => '2'
```

> how to deal with union of union types when overriding the dispatch method of the upper union type ? For example how could I use the Command() as a constructor in this case ?

The default implementation of `dispatch` should work fine.

**Comment by gcanti** [comment]
Shipped in v2.2.1 if you want to give it a spin. Thanks to all
